### PR TITLE
Removed the OpenAL/OpenGL/SDL2 dependency from the game engine

### LINKED ISSUE #5274: Renamings and Reorganisation
A collection from stuff suggested in IRC. These could be done once the pull request queue is a little less crowded https://github.com/OpenRA/OpenRA/issues/5263.

- [x] OpenRA.Support.Random clashes with System.Random https://github.com/OpenRA/OpenRA/pull/5359/
- [x] WRange → WDist ~~https://github.com/OpenRA/OpenRA/pull/4701~~ https://github.com/OpenRA/OpenRA/pull/8667
- [ ] OpenRA.Renderer.* is actually more like OpenRA.Platform.* because it is graphics + input and could also include OpenRA.Sound which is currently in OpenRA.Game so we save linking to SDL2 there. ~~https://github.com/OpenRA/OpenRA/issues/5787~~ ~~https://github.com/OpenRA/OpenRA/pull/8538~~ https://github.com/OpenRA/OpenRA/pull/8632
- [x] Many traits in OpenRA.Mods.RA should actually be in OpenRA.Mods.Common as they are shared in all three mods. Same goes for system rules every mod needs or shared chrome definitions which could also be moved to `mods/common/` for clarity. ~~https://github.com/OpenRA/OpenRA/pull/5321~~ https://github.com/OpenRA/OpenRA/pull/6511
- [x] Don't call factions countries or races. https://github.com/OpenRA/OpenRA/issues/7552

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/2079483-renamings-and-reorganisation?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by Mailaender**
I removed the OpenAL (and subsequent OpenTK) dependency from OpenRA.Game and placed it into the renderer DLLs, which in fact do a lot more than just rendering nowadays (input, window creation, etc.) so I renamed things to reflect this.

Our SDL2/OpenGL/OpenAL combination works fine, but isn't far from ideal. Now it is possible for anyone to create a DirectX platform DLL or something more light weight new based on GLFW or other stripped down OpenGL bindings etc.

Closes #5274.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -99,25 +99,25 @@ endif
...
-rnull_SRCS := $(shell find OpenRA.Renderer.Null/ -iname '*.cs')
-rnull_TARGET = OpenRA.Renderer.Null.dll
-rnull_KIND = library
-rnull_DEPS = $(game_TARGET)
-rnull_LIBS = $(COMMON_LIBS) $(rnull_DEPS)
-PROGRAMS += rsdl2 rnull
-renderers: $(rsdl2_TARGET) $(rnull_TARGET)
+# Platform dlls
+pdefault_SRCS := $(shell find OpenRA.Platforms.Default/ -iname '*.cs')
+pdefault_TARGET = OpenRA.Platforms.Default.dll
```
Comment (pchote) [inline Makefile]: `OpenRA.Platform.Default.dll` please.

Comment (Mailaender) [inline Makefile]: Nope, that will conflict with the `OpenRA.Platform` namespace and is inconsistent with `OpenRA.Mods` which is also plural.

Comment (pchote) [inline Makefile]: Okay.  I still think this is a weird naming convention, but can't fault your logic behind it.

File: OpenRA.Game/Graphics/IGraphicsDevice.cs [hunk 2]
```diff
@@ -30,6 +30,7 @@ public RendererAttribute(Type graphicsDeviceType)
 	public interface IDeviceFactory
 	{
 		IGraphicsDevice Create(Size size, WindowMode windowMode);
+		ISoundEngine Initialize();
```
Comment (pchote) [inline OpenRA.Game/Graphics/IGraphicsDevice.cs]: Can we name these consistently?  `CreateGraphics` / `CreateSound` or `InitializeGraphics` / `InitializeSound`.

File: OpenRA.Game/Renderer.cs [hunk 3]
```diff
@@ -84,6 +85,7 @@ static IGraphicsDevice CreateDevice(Assembly rendererDll, int width, int height,
 			foreach (RendererAttribute r in rendererDll.GetCustomAttributes(typeof(RendererAttribute), false))
 			{
 				var factory = (IDeviceFactory)r.Type.GetConstructor(Type.EmptyTypes).Invoke(null);
+				Sound.Engine = factory.Initialize();
```
Comment (pchote) [inline OpenRA.Game/Renderer.cs]: Tacking the audio device initialization on as a side-effect is pretty horrible.  You could either:

1.  Change this method to `static Pair<IGraphicsDevice, IAudioDevice> CreateDevices`
2.  Make it return void and set both Sound.Engine and Renderer.Device here
3.  Add a similar `CreateDevice` method to Sound and avoid mixing renderer and audio code.

My preference is strongly towards 3 (not sure why I even mention the first two, but i've already written them now...).

File: OpenRA.Game/Sound/Sound.cs [hunk 4]
```diff
@@ -16,14 +16,13 @@
...
 using OpenRA.Traits;
-using OpenTK;
-using OpenTK.Audio.OpenAL;
 
 namespace OpenRA
 {
 	public static class Sound
 	{
-		static ISoundEngine soundEngine;
+		public static ISoundEngine Engine;
```
Comment (pchote) [inline OpenRA.Game/Sound/Sound.cs]: This can remain private after you implement my suggestion above.

File: OpenRA.Game/Sound/Sound.cs [hunk 5]
```diff
@@ -48,30 +47,12 @@ static ISoundSource LoadSound(string filename)
...
-			return soundEngine.AddSoundSourceFromMemory(wave.RawOutput, wave.Channels, wave.BitsPerSample, wave.SampleRate);
+			return Engine.AddSoundSourceFromMemory(wave.RawOutput, wave.Channels, wave.BitsPerSample, wave.SampleRate);
 		}
 
 		static ISoundSource LoadSoundRaw(byte[] rawData, int channels, int sampleBits, int sampleRate)
 		{
-			return soundEngine.AddSoundSourceFromMemory(rawData, channels, sampleBits, sampleRate);
-		}
-
-		static ISoundEngine CreateEngine(string engine)
```
Comment (pchote) [inline OpenRA.Game/Sound/Sound.cs]: Ah, i'm basically just asking you to keep this, but fetch from the platform assembly instead.

**Comment by Mailaender** [comment]
I cleaned this up even further as requested.

**Comment by Mailaender** [comment]
Rebased to make sure this plays nicely with https://github.com/OpenRA/OpenRA/pull/8690.

**Comment by pchote** [comment]
Having the dll renaming, sound moving, and other code renaming all in the same commit makes this very difficult to review.  Because we are never lacking in easier prs to review, this keeps on being ignored.

Could you please break this up into smaller and more review friendly commits?

**Comment by Mailaender** [comment]
Rebased and separated the cleanup commits.

**Comment by abcdefg30** [comment]
Game still runs. :white_check_mark:

**Comment by reaperrr** [comment]
Needs rebase.

**Comment by Mailaender** [comment]
Rebased.

**Comment by Mailaender** [comment]
Rebased and resolved the per commit comments.

**Comment by penev92** [comment]
Reasoning behind this looks sound. Everything seems to work and I didn't see anything obviously wrong with the code, so :+1:

**Comment by penev92** [comment]
I feel this needs a changelog entry (removing the hardcoded dependency on SDL2/OpenGL/OpenAL and enabling the creation of a DirectX renderer sounds quite important) but I can't put the words together. Help?

File: OpenRA.Game/Sound/Sound.cs [hunk 6]
```diff
@@ -56,26 +66,12 @@ static ISoundSource LoadSoundRaw(byte[] rawData, int channels, int sampleBits, i
 			return soundEngine.AddSoundSourceFromMemory(rawData, channels, sampleBits, sampleRate);
 		}
 
-		static ISoundEngine CreateEngine(string engine)
+		public static void Initialize(SoundSettings soundSettings, ServerSettings serverSettings)
```
Comment (pchote) [inline OpenRA.Game/Sound/Sound.cs]: It's pretty bogus that our lowest level sound code cares about the server settings.

Comment (Mailaender) [inline OpenRA.Game/Sound/Sound.cs]: Your https://github.com/OpenRA/OpenRA/issues/7252 will solve this anyway.

**Comment by Mailaender** [comment]
@penev92 I am not even sure that this is changelog worthy. It is just a boring internal refactoring. Maybe the DLL renamings should be mentioned as packagers could trip over it.

**Pull request merged**